cmd/upload.go

```diff
@@ -210,6 +218,147 @@ func genStats(durations []float64) stats {
 	}
 }
 
+type uploadProgressBar struct {
+	progress           *mpb.Progress
+	bar                *mpb.Bar
+	progressTotal      int64
+	progressTotalMutex *sync.Mutex
+}
+
+func newUploadProgressBar() *uploadProgressBar {
+	progress := mpb.New()
+	var progressTotal int64
+	bar := progress.AddBar(progressTotal,
+		mpb.BarRemoveOnComplete(),
+		mpb.PrependDecorators(
+			decor.Name("upload", decor.WC{W: 7, C: decor.DidentRight}),
+			decor.OnComplete(decor.EwmaETA(decor.ET_STYLE_GO, 60, decor.WC{W: 4}), "done"),
+		),
+		mpb.AppendDecorators(decor.Percentage()),
+	)
+
+	return &uploadProgressBar{
+		progress:           progress,
+		bar:                bar,
+		progressTotal:      progressTotal,
+		progressTotalMutex: &sync.Mutex{},
+	}
+}
+
+func (upb *uploadProgressBar) incrementTotal() {
+	upb.progressTotalMutex.Lock()
+	upb.progressTotal += int64(1)
+	upb.bar.SetTotal(upb.progressTotal, false)
+	upb.progressTotalMutex.Unlock()
+}
+
+func (upb *uploadProgressBar) increment(wdd ...time.Duration) {
+	upb.bar.Increment(wdd...)
+}
+
+func (upb *uploadProgressBar) wait() {
+	upb.progress.Wait()
+}
+
+type processableFiles struct {
+	singleBundleFiles []string
+	multiBundleFiles  []string
+}
+
+func filterProcessableFiles(dir string) (processableFiles, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return processableFiles{}, err
+	}
+
+	var procFiles processableFiles
+
+	for _, file := range files {
+		if !file.IsDir() {
+			if strings.HasSuffix(file.Name(), ".json") {
+				procFiles.singleBundleFiles = append(procFiles.singleBundleFiles, filepath.Join(dir, file.Name()))
+			} else if strings.HasSuffix(file.Name(), ".ndjson") {
+				procFiles.multiBundleFiles = append(procFiles.multiBundleFiles, filepath.Join(dir, file.Name()))
+			}
+		}
+	}
+
+	return procFiles, nil
+}
+
+func enqueueBundlesFromSingleBundleFiles(files []string, uploadQueue chan<- bundle, wg *sync.WaitGroup, progressBar *uploadProgressBar) {
+	for _, file := range files {
+		progressBar.incrementTotal()
+		bundleId := bundleIdentifier{filename: file, bundleNumber: 1}
+		content, err := ioutil.ReadFile(file)
```

Comment:
I would just Open the file here and pass the Reader into the queue, the same as before. here you read the whole file in memory.

Reply:
Given that this is within the function for putting bundles into the upload queue that stem from a file only holding a single bundle, I don't quite see this as a problem. To be fair, this is under the assumption that a single bundle file will not really grow this large in size.

Your argumentation would also hold true for the other function handling multi bundle files. A single bundle in such a file could potentially be as large as one from a single bundle file. Still, it would reside in memory, too.

I can see the problem in reading from the channel and passing the content to another function which could result in a copy. Will read up on this...

Reply:
You are right, that a Bundle in a single File can be as large as Bundles in ndjson files. But opposed to ndjson files, we don't need to read the bytes into memory. Ok we could also just scan over the bytes in ndjson files and pass reader with the right offsets. 

Because the HTTP Client uses a reader in it's API, I would assume that the upload works in streaming manner. Please be aware that the largest single file which I uploaded successfully with blazectl was 37 MB. 37 MB is not small if you have to hold it in memory.

Reply:
See comment below.